src/neo4j_genai/pipeline/pipeline.py

```diff
@@ -0,0 +1,470 @@
+            )
+        if self.is_cyclic():
+            raise PipelineDefinitionError("Cyclic graph are not allowed")
+
+    def on_task_complete(self, node: TaskPipelineNode, result: RunResult) -> None:
+        res_to_save = None
+        if result.result:
+            res_to_save = result.result.model_dump()
+        self.add_result_for_component(node.name, res_to_save, is_final=node.is_leaf())
+
+    def add_result_for_component(
+        self, name: str, result: dict[str, Any] | None, is_final: bool = False
+    ) -> None:
+        self._store.add(name, result)
+        if is_final:
+            # The pipeline only returns the results
+            # of the leaf nodes
+            # TODO: make this configurable in the future.
+            self._final_results.add(name, result)
+
+    def get_results_for_component(self, name: str) -> Any:
+        return self._store.get(name)
+
+    def get_component_inputs(
+        self,
+        component_name: str,
+        input_config: dict[str, Any],
+        input_data: dict[str, Any],
+    ) -> dict[str, Any]:
+        """Find the component inputs from:
+        - data: the user input data
+        - input_config: the mapping between components results and inputs
+        """
+        component_inputs: dict[str, Any] = input_data.get(component_name, {})
+        if input_config:
+            for parameter, mapping in input_config.items():
+                try:
+                    component, output_param = mapping.split(".")
+                except ValueError:
+                    # we will use the full output of
+                    # component as input
+                    component = mapping
+                    output_param = None
+                component_result = self._store.get(component)
+                if output_param is not None:
+                    value = component_result.get(output_param)
+                else:
+                    value = component_result
+                component_inputs[parameter] = value
+        return component_inputs
+
+    def reinitialize(self) -> None:
+        """Reinitialize the result stores and component status
+        if we want to rerun the same pipeline again
+        (maybe with inputs)"""
+        self._store.empty()
+        self._final_results.empty()
+        for task in self._nodes.values():
+            task.reinitialize()
+
+    def validate_inputs_config(self, data: dict[str, Any]) -> None:
+        """Go through the graph and make sure each component will not miss any input
+
+        Args:
+            data (dict[str, Any]): the user provided data in the pipeline.run method.
+        """
+        for task in self._nodes.values():
+            task.validate_inputs_config(data)
+
+    async def run(self, data: dict[str, Any]) -> dict[str, Any]:
+        self.validate_inputs_config(data)
+        self.reinitialize()
+        orchestrator = Orchestrator(self)
+        await orchestrator.run(data)
+        return self._final_results.all()
```

Comment:
I see a use case where someone might want to have a human in the loop to verify each step before proceeding to the next one.
Do have that we'd need to support dumping config + state in a serialized format, and then hydrating a new Pipeline using the same.
And in addition to a `.run()` that runs all, have a `.step()` that takes the next step in the pipeline.

I don't see anything in this design preventing us from adding that, but I'm also looking at this code for the first time today.
So have a think if there's anything we might want to change in this PR to prevent future breaking changes when we want to add the step and dump/load functionality.

Reply:
That's a good point. For now we are tracking the state of each component to find the next one, but it is not persisted. This PR introduces a `show_as_dict` and `from_template` methods on the pipeline, that could be extended to dump/load the pipeline (and it's state). It has not been done yet because I'm unsure about how to deal with the serialization of component instances. But I don't think this would introduce breaking changes if we want to add this feature in the future.